# Advent of Code, 11th Dec 2024

## Plutonian Pebbles - Part 1

The ancient civilization on Pluto was known for its ability to manipulate spacetime, and while The Historians explore their infinite corridors, you've noticed a strange set of physics-defying stones.

At first glance, they seem like normal stones: they're arranged in a perfectly **straight line**, and each stone has a **number** engraved on it.

The strange part is that every time you blink, the stones **change**.

Sometimes, the number engraved on a stone changes. Other times, a stone might **split in two**, causing all the other stones to shift over a bit to make room in their perfectly straight line.

As you observe them for a while, you find that the stones have a consistent behavior. Every time you blink, the stones each **simultaneously change** according to the **first applicable rule** in this list:

* If the stone is engraved with the number 0, it is replaced by a stone engraved with the number 1.
* If the stone is engraved with a number that has an **even** number of digits, it is replaced by **two stones**. The left half of the digits are engraved on the new left stone, and the right half of the digits are engraved on the new right stone. (The new numbers don't keep extra leading zeroes: 1000 would become stones 10 and 0.)
* If none of the other rules apply, the stone is replaced by a new stone; the old stone's number **multiplied by 2024** is engraved on the new stone.

No matter how the stones change, their order is preserved, and they stay on their perfectly straight line.

How will the stones evolve if you keep blinking at them? You take a note of the number engraved on each stone in the line (your puzzle input).

If you have an arrangement of five stones engraved with the numbers 0 1 10 99 999 and you blink once, the stones transform as follows:

* The first stone, 0, becomes a stone marked 1.
* The second stone, 1, is multiplied by 2024 to become 2024.
* The third stone, 10, is split into a stone marked 1 followed by a stone marked 0.
* The fourth stone, 99, is split into two stones marked 9.
* The fifth stone, 999, is replaced by a stone marked 2021976.

So, after blinking once, your five stones would become an arrangement of seven stones engraved with the numbers 
```
1 2024 1 0 9 9 2021976.
```

Here is a longer example:

Initial arrangement:
```
125 17
```

After 1 blink:
```
253000 1 7
```

After 2 blinks:
```
253 0 2024 14168
```

After 3 blinks:
```
512072 1 20 24 28676032
```

After 4 blinks:
```
512 72 2024 2 0 2 4 2867 6032
```

After 5 blinks:
```
1036288 7 2 20 24 4048 1 4048 8096 28 67 60 32
```

After 6 blinks:
```
2097446912 14168 4048 2 0 2 4 40 48 2024 40 48 80 96 2 8 6 7 6 0 3 2
```

In this example, after blinking six times, you would have 22 stones. After blinking 25 times, you would have **55312** stones!

Consider the arrangement of stones in front of you. **How many stones will you have after blinking 25 times?**

### Read in the data

In [18]:
#!F#
let inputFilePath = "aoc-2024-12-11-puzzle-input.txt"

Read the lines from the file.

In [19]:
open System.IO

let lines: string list = File.ReadAllLines(inputFilePath) |> Array.toList
lines.Length

1

In [20]:
let stoneNumbers = lines[0].Split(" ") |> Array.map Int64.Parse |> Array.toList
stoneNumbers

[ 5688, 62084, 2, 3248809, 179, 79, 0, 172169 ] HeadOrDefault 5688 TailOrNull [ 62084, 2, 3248809, 179, 79, 0, 172169 ] HeadOrDefault 62084 TailOrNull [ 2, 3248809, 179, 79, 0, 172169 ] HeadOrDefault 2 TailOrNull [ 3248809, 179, 79, 0, 172169 ] HeadOrDefault 3248809 TailOrNull [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] (values) [ 179, 79, 0, 172169 ] Head 3248809 Tail [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] (values) [ 179, 79, 0, 172169 ] (values) [ 3248809, 179, 79, 0, 172169 ] Head 2 Tail [ 3248809, 179, 79, 0, 172169 ] HeadOrDefault 3248809 TailOrNull [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] (values) [ 179, 79, 0, 172169 ] Head 3248809 Tail [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] (values) [ 179, 79, 0, 172169 ] (values) [ 3248809, 179, 79, 0, 172169 ] (values) [ 2, 3248809, 179, 79, 0, 172169 ] Head 62084 Tail [ 2, 3248809, 179, 79, 0, 172169 ] HeadOrDefault 2 TailOrNull [ 3248809, 179, 79, 0, 172169 ] HeadOrDefault 3248809 TailOrNull [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] (values) [ 179, 79, 0, 172169 ] Head 3248809 Tail [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] (values) [ 179, 79, 0, 172169 ] (values) [ 3248809, 179, 79, 0, 172169 ] Head 2 Tail [ 3248809, 179, 79, 0, 172169 ] HeadOrDefault 3248809 TailOrNull [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] (values) [ 179, 79, 0, 172169 ] Head 3248809 Tail [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] (values) [ 179, 79, 0, 172169 ] (values) [ 3248809, 179, 79, 0, 172169 ] (values) [ 2, 3248809, 179, 79, 0, 172169 ] (values) [ 62084, 2, 3248809, 179, 79, 0, 172169 ] Head 5688 Tail [ 62084, 2, 3248809, 179, 79, 0, 172169 ] HeadOrDefault 62084 TailOrNull [ 2, 3248809, 179, 79, 0, 172169 ] HeadOrDefault 2 TailOrNull [ 3248809, 179, 79, 0, 172169 ] HeadOrDefault 3248809 TailOrNull [ 179, 79, 0, 172169 ] HeadOrDefault 179 TailOrNull [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [ 0, 172169 ] Head 79 Tail [ 0, 172169 ] (values) [ 79, 0, 172169 ] Head 179 Tail [ 79, 0, 172169 ] HeadOrDefault 79 TailOrNull [

### Solve the problem

In [21]:
let applyStoneRules (stoneNumbers: int64 list): int64 list = (seq {
    for stoneNumber in stoneNumbers do
        let numberStr = stoneNumber.ToString()
        match stoneNumber with
        | 0L -> yield 1L
        | evenish when numberStr.Length % 2 = 0 ->
            let halfLength = numberStr.Length / 2
            let firstHalf = numberStr[0..(halfLength-1)]
            let secondHalf = numberStr[halfLength..]
            let firstHalfInt = Int32.Parse(firstHalf)
            let secondHalfInt = Int32.Parse(secondHalf)
            yield firstHalfInt
            yield secondHalfInt
        | n -> yield n * 2024L
} |> Seq.toList)

In [22]:
let rec iterate (n: int) (f: 'a -> 'a) (x: 'a): 'a =
    if n = 0 then x
    else iterate (n-1) f (f x)

In [26]:
let newNumbers = stoneNumbers |> iterate 25 applyStoneRules
newNumbers

[ 24, 57, 94, 56, 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144 ... (more) ] HeadOrDefault 24 TailOrNull [ 57, 94, 56, 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2 ... (more) ] HeadOrDefault 57 TailOrNull [ 94, 56, 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0 ... (more) ] HeadOrDefault 94 TailOrNull [ 56, 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2 ... (more) ] HeadOrDefault 56 TailOrNull [ 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4 ... (more) ] HeadOrDefault 2 TailOrNull [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] HeadOrDefault 0 TailOrNull [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] Head 0 Tail [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] (values) [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] Head 2 Tail [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] HeadOrDefault 0 TailOrNull [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] Head 0 Tail [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] (values) [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] (values) [ 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4 ... (more) ] Head 56 Tail [ 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4 ... (more) ] HeadOrDefault 2 TailOrNull [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] HeadOrDefault 0 TailOrNull [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] Head 0 Tail [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] (values) [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] Head 2 Tail [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] HeadOrDefault 0 TailOrNull [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] Head 0 Tail [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] (values) [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] (values) [ 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4 ... (more) ] (values) [ 56, 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2 ... (more) ] Head 94 Tail [ 56, 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2 ... (more) ] HeadOrDefault 56 TailOrNull [ 2, 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4 ... (more) ] HeadOrDefault 2 TailOrNull [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] HeadOrDefault 0 TailOrNull [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] Head 0 Tail [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] (values) [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] Head 2 Tail [ 0, 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192 ... (more) ] HeadOrDefault 0 TailOrNull [ 2, 4, 28, 67, 60, 32, 8096, 1, 8096, 16192, 16192, 1, 18216, 12144, 2, 0, 2, 4, 16192, 1 ... (more) ] Head 0 Tail [ 2, 4, 28,

In [27]:
newNumbers.Length

186175

The above is the answer that we need to Part 1.

In [28]:
newNumbers |> Set.ofList |> Set.count

522

Very interestingly, of 186175 numbers there are only 522 unique numbers, which means that we have a means to greatly compress both the memory usage and calculation time.

## Plutonian Pebbles - Part 2

The Historians sure are taking a long time. To be fair, the infinite corridors **are** very large.

**How many stones would you have after blinking a total of 75 times?**

### Solve the problem

In [25]:
// let mutable totalCount40 = 0L

// let rec countStoneNumbers (iterations: int) (startNumber: int64): int64 =
//     if iterations = 0 then 1L
//     else
//         let newNumbers = [startNumber] |> applyStoneRules
//         let newCounts =
//             if (true) then // iterations <= 40
//                 newNumbers |> List.map (fun number -> countStoneNumbers (iterations - 1) number)
//             else
//                 newNumbers
//                 |> Array.ofList
//                 |> Array.Parallel.map (fun number -> countStoneNumbers (iterations - 1) number)
//                 |> Array.toList
//         let result = newCounts |> List.sum
//         result

// let rec countStoneNumbersList (iterations: int) (startNumbers: int64 list): int64 =
//     for idx in 0..iterations do
//         iterationCounts[idx] <- 0L
//     let newCounts = startNumbers |> List.map (fun number -> countStoneNumbers iterations number)
//     newCounts |> List.sum

In [34]:
type ValueCount = int64 * int64

[<TailCall>] // tail call optimization
let rec iterate2 (n: int) (f: int64 -> int64 seq) (x: ValueCount list): ValueCount list =
    if n = 0 then x
    else
        let newRawValueCountsUngroupedSeq = seq {
            for (value, count) in x do
                yield! (f value) |> Seq.map (fun newValue -> (newValue, count))
        }
        let newValueCountGroupedList = newRawValueCountsUngroupedSeq |> Seq.groupBy fst |> Seq.map (fun (key, values) -> (key, values |> Seq.sumBy snd)) |> Seq.toList
        iterate2 (n-1) f newValueCountGroupedList

In [35]:
let applyStoneRules2 (stoneNumber: int64): int64 seq = seq {
    let numberStr = stoneNumber.ToString()
    match stoneNumber with
    | 0L -> yield 1L
    | evenish when numberStr.Length % 2 = 0 ->
        let halfLength = numberStr.Length / 2
        let firstHalf = numberStr[0..(halfLength-1)]
        let secondHalf = numberStr[halfLength..]
        let firstHalfInt = Int32.Parse(firstHalf)
        let secondHalfInt = Int32.Parse(secondHalf)
        yield firstHalfInt
        yield secondHalfInt
    | n -> yield n * 2024L
}

In [37]:
let newValuesCounts = stoneNumbers |> List.map (fun number -> (number, 1L)) |> iterate2 25 applyStoneRules2
newValuesCounts

[ (24, 4497), (57, 1502), (94, 1504), (56, 1503), (2, 10202), (0, 10101), (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835) ... (more) ] HeadOrDefault (24, 4497) Item1 24 Item2 4497 TailOrNull [ (57, 1502), (94, 1504), (56, 1503), (2, 10202), (0, 10101), (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250) ... (more) ] HeadOrDefault (57, 1502) Item1 57 Item2 1502 TailOrNull [ (94, 1504), (56, 1503), (2, 10202), (0, 10101), (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008) ... (more) ] HeadOrDefault (94, 1504) Item1 94 Item2 1504 TailOrNull [ (56, 1503), (2, 10202), (0, 10101), (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008), (24579456, 3203) ... (more) ] HeadOrDefault (56, 1503) Item1 56 Item2 1503 TailOrNull [ (2, 10202), (0, 10101), (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008), (24579456, 3203), (28676032, 1836) ... (more) ] HeadOrDefault (2, 10202) Item1 2 Item2 10202 TailOrNull [ (0, 10101), (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008), (24579456, 3203), (28676032, 1836), (2024, 4192) ... (more) ] HeadOrDefault (0, 10101) TailOrNull [ (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008), (24579456, 3203), (28676032, 1836), (2024, 4192), (72, 918) ... (more) ] Head (0, 10101) Tail [ (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008), (24579456, 3203), (28676032, 1836), (2024, 4192), (72, 918) ... (more) ] (values) index value 0 (0, 10101) 1 (4, 11120) 2 (28, 1287) 3 (67, 869) 4 (60, 1787) 5 (32, 2894) 6 (8096, 7051) 7 (1, 7103) 8 (16192, 6885) 9 (18216, 3041) 10 (12144, 5464) 11 (77, 2023) 12 (26, 2022) 13 (8, 10738) 14 (40, 3835) 15 (48, 4250) 16 (32772608, 4008) 17 (24579456, 3203) 18 (28676032, 1836) 19 (2024, 4192) ... (more) Head (2, 10202) Item1 2 Item2 10202 Tail [ (0, 10101), (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008), (24579456, 3203), (28676032, 1836), (2024, 4192) ... (more) ] HeadOrDefault (0, 10101) TailOrNull [ (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008), (24579456, 3203), (28676032, 1836), (2024, 4192), (72, 918) ... (more) ] Head (0, 10101) Tail [ (4, 11120), (28, 1287), (67, 869), (60, 1787), (32, 2894), (8096, 7051), (1, 7103), (16192, 6885), (18216, 3041), (12144, 5464), (77, 2023), (26, 2022), (8, 10738), (40, 3835), (48, 4250), (32772608, 4008), (24579456, 3203), (28676032, 1836), (2024, 4192), (72, 918) ... (more) ] (values) index value 0 (0, 10101) 1 (4, 11120) 2 (28, 1287) 3 (67, 869) 4 (60, 1787) 5 (32, 2894) 6 (8096, 7051) 7 (1, 7103) 8 (16192, 6885) 9 (18216, 3041) 10 (12144, 5464) 11 (77, 2023) 1

In [38]:
newValuesCounts |> List.map (fun (value, count) -> count) |> List.sum

186175

OK, now lets try with 75 iterations.

In [48]:
let newValuesCounts2 = stoneNumbers |> List.map (fun number -> (number, 1L)) |> iterate2 75 applyStoneRules2
newValuesCounts2

[ (3, 2895871842033), (2, 11771661536922), (7, 5777200288148), (6, 9896160731120), (16192, 8028914145549), (4048, 7753721490714), (1, 8189541757949), (8096, 8224318764053), (8, 12183357243504), (9, 5459656371704), (4, 12478376331087), (5, 2447103486281), (0, 11402625247625), (18216, 3600452525107), (12144, 6526264968316), (2867, 1648501101825), (6032, 1648501101956), (2457, 2824744047509), (9456, 2824744047509), (32, 3379844699180) ... (more) ] HeadOrDefault (3, 2895871842033) Item1 3 Item2 2895871842033 TailOrNull [ (2, 11771661536922), (7, 5777200288148), (6, 9896160731120), (16192, 8028914145549), (4048, 7753721490714), (1, 8189541757949), (8096, 8224318764053), (8, 12183357243504), (9, 5459656371704), (4, 12478376331087), (5, 2447103486281), (0, 11402625247625), (18216, 3600452525107), (12144, 6526264968316), (2867, 1648501101825), (6032, 1648501101956), (2457, 2824744047509), (9456, 2824744047509), (32, 3379844699180), (77, 2291830091894) ... (more) ] HeadOrDefault (2, 11771661536922) Item1 2 Item2 11771661536922 TailOrNull [ (7, 5777200288148), (6, 9896160731120), (16192, 8028914145549), (4048, 7753721490714), (1, 8189541757949), (8096, 8224318764053), (8, 12183357243504), (9, 5459656371704), (4, 12478376331087), (5, 2447103486281), (0, 11402625247625), (18216, 3600452525107), (12144, 6526264968316), (2867, 1648501101825), (6032, 1648501101956), (2457, 2824744047509), (9456, 2824744047509), (32, 3379844699180), (77, 2291830091894), (26, 2291830090875) ... (more) ] HeadOrDefault (7, 5777200288148) Item1 7 Item2 5777200288148 TailOrNull [ (6, 9896160731120), (16192, 8028914145549), (4048, 7753721490714), (1, 8189541757949), (8096, 8224318764053), (8, 12183357243504), (9, 5459656371704), (4, 12478376331087), (5, 2447103486281), (0, 11402625247625), (18216, 3600452525107), (12144, 6526264968316), (2867, 1648501101825), (6032, 1648501101956), (2457, 2824744047509), (9456, 2824744047509), (32, 3379844699180), (77, 2291830091894), (26, 2291830090875), (3277, 3477211845040) ... (more) ] HeadOrDefault (6, 9896160731120) Item1 6 Item2 9896160731120 TailOrNull [ (16192, 8028914145549), (4048, 7753721490714), (1, 8189541757949), (8096, 8224318764053), (8, 12183357243504), (9, 5459656371704), (4, 12478376331087), (5, 2447103486281), (0, 11402625247625), (18216, 3600452525107), (12144, 6526264968316), (2867, 1648501101825), (6032, 1648501101956), (2457, 2824744047509), (9456, 2824744047509), (32, 3379844699180), (77, 2291830091894), (26, 2291830090875), (3277, 3477211845040), (2608, 3477211845053) ... (more) ] HeadOrDefault (16192, 8028914145549) Item1 16192 Item2 8028914145549 TailOrNull [ (4048, 7753721490714), (1, 8189541757949), (8096, 8224318764053), (8, 12183357243504), (9, 5459656371704), (4, 12478376331087), (5, 2447103486281), (0, 11402625247625), (18216, 3600452525107), (12144, 6526264968316), (2867, 1648501101825), (6032, 1648501101956), (2457, 2824744047509), (9456, 2824744047509), (32, 3379844699180), (77, 2291830091894), (26, 2291830090875), (3277, 3477211845040), (2608, 3477211845053), (3686, 1558810323842) ... (more) ] HeadOrDefault (4048, 7753721490714) TailOrNull [ (1, 8189541757949), (8096, 8224318764053), (8, 12183357243504), (9, 5459656371704), (4, 12478376331087), (5, 2447103486281), (0, 11402625247625), (18216, 3600452525107), (12144, 6526264968316), (2867, 1648501101825), (6032, 1648501101956), (2457, 2824744047509), (9456, 2824744047509), (32, 3379844699180), (77, 2291830091894), (26, 2291830090875), (3277, 3477211845040), (2608, 3477211845053), (3686, 1558810323842), (9184, 1558810323842) ... (more) ] Head (4048, 7753721490714) Tail [ (1, 8189541757949), (8096, 8224318764053), (8, 12183357243504), (9, 5459656371704), (4, 12478376331087), (5, 2447103486281), (0, 11402625247625), (18216, 3600452525107), (12144, 6526264968316), (2867, 1648501101825), (6032, 1648501101956), (2457, 2824744047509), (9456, 2824744047509), (32, 3379844699180), (77, 2291830091894), (26, 2291830090875), (3277, 3477211845040), (2608, 34

In [49]:
newValuesCounts2 |> List.map (fun (value, count) -> count) |> List.sum

220566831337810

And this is the right answer!  It actually took a lot of time to get the code down to something this simple, it took a lot of time to realise that in the list of stone numbers, the number of unique values was so very small.